### PR TITLE
test: Add test for wtxid transaction relay

### PR DESCRIPTION
**Pull request by fjahr**
Based on #18044 

This updates the functional testing framework to the p2p protocol version including wtxid transaction relay (70016) and adds a basic test for wtxid based relay.

### CODE REVIEW

File: test/functional/test_framework/messages.py [hunk 1]
```diff
@@ -31,7 +31,7 @@
 from test_framework.util import hex_str_to_bytes, assert_equal
 
 MIN_VERSION_SUPPORTED = 60001
-MY_VERSION = 70014  # past bip-31 for ping/pong
+MY_VERSION = 70016  # past bip-31 for ping/pong
```
Comment (sipa) [inline test/functional/test_framework/messages.py]: Comment is outdated now.

**Comment by ajtowns** [comment]
I think this isn't testing:

 * whether the node switches from sending TX invs to WTX invs when the wtxidrelay message is sent
 * whether we get a wtxidrelay message when we give a 70016 version, but not when we give a lower version
 * in either case, when we announce a TX inv, do we get a TX getdata back, and likewise for WTX?
 * when a tx is sent (not an INV but the actual tx) if it's an orphan, the missing parent's txid is requested as a TX not a WTX

I've had a go at some of the changes: https://github.com/ajtowns/bitcoin/tree/202003-wtxidrelay-test but haven't made it so it actually sends the `wtxidrelay` message.

**Comment by fjahr** [comment]
@ajtowns Thank you, that is very helpful. My intuition was that many of these cases would be tested indirectly by upgrading the test framework but of course, explicit tests are always better and, as I have realized now, I also had not implemented it correctly. Your tests mainly did not pass because I had not implemented feature negotiation correctly. This is done now and I preserved your commit mostly so it is easier for you to review. If the test now does what you intended it to do then I can squash.

Overview of changes:
- Rebased
- Fixed outdated comment
- Added explicit test for wtxidrelay
- Fixed wtxidrelay feature negotiation
- Added @ajtowns code and with some further changes on top

File: test/functional/test_framework/messages.py (hunk 1 above)
Comment (fjahr) [reply test/functional/test_framework/messages.py]: fixed

**Review by ajtowns — CHANGES_REQUESTED** [review]
I think this needs more changes to match the latest update to #18044; I merged them and added a fixup commit on https://github.com/ajtowns/bitcoin/tree/202003-wtxidrelay-test if you want to take a look. Feel free to squash the fixups anytime btw.

File: test/functional/p2p_segwit.py [hunk 2]
```diff
@@ -143,16 +146,24 @@ def test_witness_block(node, p2p, block, accepted, with_witness=True, reason=Non
...
     # Avoid sending out msg_getdata in the mininode thread as a reply to invs.
     # They are not needed and would only lead to races because we send msg_getdata out in the test thread
     def on_inv(self, message):
         pass
 
+    def on_version(self, message):
+        if self.wtxidrelay:
+            self.send_message(msg_wtxidrelay())
+        self.send_message(msg_verack())
+        self.nServices = message.nServices
```
[the remark this answers is not in the document]
Comment (fjahr) [reply test/functional/p2p_segwit.py]: done

**Comment by fjahr** [comment]
Thanks @ajtowns ! I pulled in latest changes in #18044 and your suggestions. Then squashed and edited commit messages for easier reviewing. Also left your suggested change `ignore non-wtxidrelay compliant invs` in for now, pending discussion in #18044 .

**Comment by fjahr** [comment]
Pushed an update that should make all functional tests use wtxid relay by default (Github doesn't seem to be working properly at the moment, here is the commit: https://github.com/fjahr/bitcoin/commit/becdb68ab37fdcbc1dcf743b19ea8629f52677bb)

**Comment by sdaftuar** [comment]
FYI -- I tried to cherry-pick the test commits here onto the latest version of #18044 but the p2p_segwit.py test is failing for me.

**Comment by fjahr** [comment]
> FYI -- I tried to cherry-pick the test commits here onto the latest version of #18044 but the p2p_segwit.py test is failing for me.

@sdaftuar Thanks for the heads-up! I rebased with the latest state of #18044 and made several fixes plus squashed to three commits now.

**Comment by fjahr** [comment]
These commits were pulled into #18044 so it can be closed.

**Pull request closed without merging**